File: ios/Sources/DBRPlugin/DBRPlugin.swift
import Foundation
import Capacitor
import DynamsoftCore
import DynamsoftLicense
import DynamsoftCaptureVisionRouter
import DynamsoftBarcodeReader

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(DBRPlugin)
public class DBRPlugin: CAPPlugin, LicenseVerificationListener   {
    private var cvr:CaptureVisionRouter!;
    private var licenseCall:CAPPluginCall!;
    @objc func initialize(_ call: CAPPluginCall) {
        if cvr == nil {
            cvr = CaptureVisionRouter()
        }
        call.resolve()
    }
    
    public func onLicenseVerified(_ isSuccess: Bool, error: Error?) {
        if isSuccess {
            licenseCall.resolve()
        }else{
            licenseCall.reject(error?.localizedDescription ?? "license error")
        }
        licenseCall = nil
   }

    
    @objc func initLicense(_ call: CAPPluginCall) {
        call.keepAlive = true
        licenseCall = call
        var license = call.getString("license") ?? "DLS2eyJoYW5kc2hha2VDb2RlIjoiMjAwMDAxLTE2NDk4Mjk3OTI2MzUiLCJvcmdhbml6YXRpb25JRCI6IjIwMDAwMSIsInNlc3Npb25QYXNzd29yZCI6IndTcGR6Vm05WDJrcEQ5YUoifQ=="
        LicenseManager.initLicense(license, verificationDelegate: self)
    }
    
    @objc func initRuntimeSettingsFromString(_ call: CAPPluginCall) {
        let template = call.getString("template") ?? ""
        if cvr != nil {
            if template != "" {
                do {
                    try cvr.initSettings(template)
                    call.resolve()
                }catch {
                    print("Unexpected error: \(error).")
                    call.reject(error.localizedDescription)
                }
            }else{
                call.reject("Empty template")
            }
        }else{
            call.reject("DDN not initialized")
        }
    }
    
    @objc func decode(_ call: CAPPluginCall) {
        var returned_results: [Any] = []
        let template = call.getString("template") ?? PresetTemplate.readBarcodes.rawValue
        var capturedResult:CapturedResult
        let path = call.getString("path") ?? ""
        var base64 = call.getString("source") ?? ""
        if path != "" {
            capturedResult = cvr.captureFromFile(path, templateName: template)
        }else{
            base64 = Utils.removeDataURLHead(base64)
            let image = Utils.convertBase64ToImage(base64)
            capturedResult = cvr.captureFromImage(image!, templateName: template)
        }
        let results = capturedResult.items
        if results != nil {
            for result in results! {
                returned_results.append(Utils.wrapBarcodeResult(result:result as! BarcodeResultItem))
            }
        }
        call.resolve(["results":returned_results])
    }
    
    @objc func decodeBitmap(_ call: CAPPluginCall) {
        let template = call.getString("template") ?? PresetTemplate.readBarcodes.rawValue
        let interop = Interoperator()
        let className = call.getString("className") ?? "CameraPreviewPlugin"
        let methodName = call.getString("methodName") ?? "getBitmap"
        let image = interop.getUIImage(className,methodName: methodName)
        var returned_results: [Any] = []
        if image != nil {
            let capturedResult = cvr.captureFromImage(image!, templateName: template)
            let results = capturedResult.items
            if results != nil {
                for result in results! {
                    returned_results.append(Utils.wrapBarcodeResult(result:result as! BarcodeResultItem))
                }
            }
        }
        call.resolve(["results":returned_results])
    }
}

File: android/src/main/java/com/dynamsoft/capacitor/dbr/DBRPlugin.java
package com.dynamsoft.capacitor.dbr;

import android.graphics.Bitmap;
import android.util.Log;

import com.dynamsoft.core.basic_structures.CapturedResultItem;
import com.dynamsoft.cvr.CaptureVisionRouter;
import com.dynamsoft.cvr.CaptureVisionRouterException;
import com.dynamsoft.cvr.CapturedResult;
import com.dynamsoft.cvr.EnumPresetTemplate;
import com.dynamsoft.dbr.BarcodeResultItem;
import com.dynamsoft.license.LicenseManager;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

@CapacitorPlugin(name = "DBR")
public class DBRPlugin extends Plugin {
    private CaptureVisionRouter cvr;
    @PluginMethod
    public void initialize(PluginCall call) {
        if (cvr == null) {
            cvr = new CaptureVisionRouter(getContext());
        }
        call.resolve();
    }

    @PluginMethod
    public void initLicense(PluginCall call) {
        String license = call.getString("license","DLS2eyJoYW5kc2hha2VDb2RlIjoiMjAwMDAxLTE2NDk4Mjk3OTI2MzUiLCJvcmdhbml6YXRpb25JRCI6IjIwMDAwMSIsInNlc3Npb25QYXNzd29yZCI6IndTcGR6Vm05WDJrcEQ5YUoifQ==");
        LicenseManager.initLicense(license, getContext(), (isSuccess, error) -> {
            if (!isSuccess) {
                Log.e("DBR", "InitLicense Error: " + error);
                call.reject(error.getMessage());
            }else{
                Log.d("DBR","license valid");
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void initRuntimeSettingsFromString(PluginCall call) {
        String template = call.getString("template");
        if (cvr != null) {
            try {
                cvr.initSettings(template);
                call.resolve();
            } catch (CaptureVisionRouterException e) {
                e.printStackTrace();
                call.reject(e.getMessage());
            }
        }else{
            call.reject("DBR not initialized");
        }
    }

    @PluginMethod
    public void decode(PluginCall call) {
        String path = call.getString("path","");
        String source = call.getString("source","");
        String templateName = call.getString("template", EnumPresetTemplate.PT_READ_BARCODES);
        if (path.equals("")) {
            source = source.replaceFirst("data:.*?;base64,","");
        }
        if (cvr != null) {
            try {
                JSObject response = new JSObject();
                JSArray barcodeResults = new JSArray();
                CapturedResult capturedResult;
                if (path.equals("")) {
                    capturedResult = cvr.capture(Utils.base642Bitmap(source),templateName);
                }else{
                    capturedResult = cvr.capture(Utils.base642Bitmap(path),templateName);
                }
                CapturedResultItem[] results = capturedResult.getItems();
                if (results != null) {
                    for (CapturedResultItem result:results) {
                        if (result instanceof BarcodeResultItem) {
                            barcodeResults.put(Utils.getMapFromBarcodeResultItem((BarcodeResultItem) result));
                        }
                    }
                }
                response.put("results",barcodeResults);
                call.resolve(response);
            } catch (Exception e) {
                e.printStackTrace();
                call.reject(e.getMessage());
            }
        }else{
            call.reject("DBR not initialized");
        }
    }

    @PluginMethod
    public void decodeBitmap(PluginCall call) {
        if (cvr != null) {
            try {
                JSObject response = new JSObject();
                JSArray barcodeResults = new JSArray();
                String templateName = call.getString("template",EnumPresetTemplate.PT_READ_BARCODES);
                String className = call.getString("className","com.tonyxlh.capacitor.camera.CameraPreviewPlugin");
                String methodName = call.getString("methodName","getBitmap");
                Class cls = Class.forName(className);
                Method m = cls.getMethod(methodName,null);
                Bitmap bitmap = (Bitmap) m.invoke(null, null);
                if (bitmap != null) {
                    CapturedResult capturedResult = cvr.capture(bitmap,templateName);
                    CapturedResultItem[] results = capturedResult.getItems();
                    if (results != null) {
                        for (CapturedResultItem result:results) {
                            if (result instanceof BarcodeResultItem) {
                                barcodeResults.put(Utils.getMapFromBarcodeResultItem((BarcodeResultItem) result));
                            }
                        }
                    }
                }
                response.put("results",barcodeResults);
                call.resolve(response);
            } catch (ClassNotFoundException | NoSuchMethodException | IllegalAccessException |
                     InvocationTargetException e) {
                e.printStackTrace();
                call.reject(e.getMessage());
            }
        }else{
            call.reject("DBR not initialized");
        }
    }
}